vendor/github.com/mongodb/mongo-tools-common/db/buffered_bulk.go

```diff
@@ -12,85 +12,99 @@ import (
 	bb := &BufferedBulkInserter{
-		collection:      collection,
-		unordered:       unordered,
-		docLimit:        docLimit,
+		collection:    collection,
+		bulkWriteOpts: options.BulkWrite().SetOrdered(!unordered),
+		docLimit:      docLimit,
 	}
 	bb.resetBulk()
 	return bb
 }
 
 func (bb *BufferedBulkInserter) Unordered() {
-	bb.unordered = true
+	bb.bulkWriteOpts.SetOrdered(false)
 }
 
 func (bb *BufferedBulkInserter) SetBypassDocumentValidation(bypass bool) {
-	bb.bypassDocumentValidation = bypass
+	bb.bulkWriteOpts.SetBypassDocumentValidation(bypass)
+}
+
+func (bb *BufferedBulkInserter) SetUpsert(upsert bool) {
+	bb.upsert = upsert
 }
 
 // throw away the old bulk and init a new one
 func (bb *BufferedBulkInserter) resetBulk() {
-	bb.documents = make([]interface{}, 0, bb.docLimit)
-	bb.byteCount = 0
+	bb.writeModels = make([]mongo.WriteModel, 0, bb.docLimit)
 	bb.docCount = 0
 }
 
-// Insert adds a document to the buffer for bulk insertion. If the buffer is
-// full, the bulk insert is made, returning any error that occurs.
+// Insert adds a document to the buffer for bulk insertion. If the buffer is full, the bulk write is done, returning
+// any error that occurs.
 func (bb *BufferedBulkInserter) Insert(doc interface{}) error {
 	rawBytes, err := bson.Marshal(doc)
 	if err != nil {
-		return fmt.Errorf("bson encoding error: %v", err)
+		return fmt.Errorf("bson encoding err: %v", err)
 	}
 
 	return bb.InsertRaw(rawBytes)
 }
 
+// Update adds a document to the buffer for bulk update. If the buffer is full, the bulk write is done, returning
+// any error that occurs.
+func (bb *BufferedBulkInserter) Update(selector, update bson.D) error {
+	return bb.addModel(mongo.NewUpdateOneModel().SetFilter(selector).SetUpdate(update).SetUpsert(bb.upsert))
+}
+
+// Replace adds a document to the buffer for bulk replacement. If the buffer is full, the bulk write is done, returning
+// any error that occurs.
+func (bb *BufferedBulkInserter) Replace(selector, replacement bson.D) error {
+	return bb.addModel(mongo.NewReplaceOneModel().SetFilter(selector).SetReplacement(replacement).SetUpsert(bb.upsert))
+}
+
 // InsertRaw adds a document, represented as raw bson bytes, to the buffer for bulk insertion. If the buffer is full,
-// the bulk insert is made, returning any error that occurs.
-func (bb *BufferedBulkInserter) InsertRaw(rawBytes []byte) (err error) {
-	// flush if we are full
-	//
-	// XXX With OP_MSG the limit is larger; MaxBSONSize shouldn't be hard
-	// coded, it should be based on the server's ismaster response.
-	if bb.docCount >= bb.docLimit || bb.byteCount+len(rawBytes) > MaxBSONSize {
+// the bulk write is done, returning any error that occurs.
+func (bb *BufferedBulkInserter) InsertRaw(rawBytes []byte) error {
+	return bb.addModel(mongo.NewInsertOneModel().SetDocument(rawBytes))
+}
+
+func (bb *BufferedBulkInserter) addModel(model mongo.WriteModel) error {
+	var err error
+	if bb.docCount > bb.docLimit {
```

Comment:
I think it would be better if we make this check (and subsequent flush) _after_ adding the model. That way, the write occurs as soon as the buffer is full, rather than waiting for the next insert or an explicit flush.

Reply:
done